### PR TITLE
refactor(core): decrease number of times model definitions are built

### PR DESCRIPTION
**Pull request by craicoverflow**
This splits the `models` metadata variable into a getter/setter to prevent building all models every time the `GraphbackCoreMetadata#getModelDefinitions` is called.

I ran benchmarks a few times and there is an improvement in startup time:

```
│ Server                                        │ Requests/s │ Latency │ Throughput/Mb │ Startup time (s) │ Resident Set Size (MB) │ CPU Percentage │
│ 1.0.0-graphback-apollo-express-large-schema   │ 256.4      │ 18.97   │ 3.59          │ 5.71             │ 134.94                 │ 1.51           │
│ 0.16.2-graphback-apollo-express-large-schema  │ 252.2      │ 19.27   │ 3.53          │ 6.08             │ 135.13                 │ 1.58           │
```

### CODE REVIEW

File: packages/graphback-core/src/plugin/GraphbackCoreMetadata.ts
Comment (machi1990) [inline packages/graphback-core/src/plugin/GraphbackCoreMetadata.ts:68]: I'd expect a setter method to receive models as parameters but here we are not doing that but instead we are parsing/computing model definitions. can we rename the method in that regard?

File: packages/graphback-core/src/plugin/GraphbackPluginEngine.ts
Comment (machi1990) [inline packages/graphback-core/src/plugin/GraphbackPluginEngine.ts:79]:
```suggestion
    for (const plugin of this.plugins) { 
      this.metadata.setModelDefinitions();
      const newSchema = plugin.transformSchema(this.metadata);
      this.metadata.setSchema(newSchema);
    }
```

File: packages/graphback-runtime-knex/tests/data/KnexDbDataProviderTest.ts
Comment (machi1990) [inline packages/graphback-runtime-knex/tests/data/KnexDbDataProviderTest.ts:8]: can we make `graphback-codegen-schema` dev dependency and use that here instead of relative path?

File: packages/graphback-runtime-knex/tests/service/DefaultCRUDServiceTest.ts
Comment (machi1990) [inline packages/graphback-runtime-knex/tests/service/DefaultCRUDServiceTest.ts:10]: See https://github.com/aerogear/graphback/pull/2057/files#r489394668

File: packages/graphback-runtime-mongodb/tests/MongoDataProviderTest.ts
Comment (machi1990) [inline packages/graphback-runtime-mongodb/tests/MongoDataProviderTest.ts:5]: See https://github.com/aerogear/graphback/pull/2057/files#r489394668

File: packages/graphback-runtime-mongodb/tests/__util__.ts
Comment (machi1990) [inline packages/graphback-runtime-mongodb/tests/__util__.ts:5]: See https://github.com/aerogear/graphback/pull/2057/files#r489394668

**Review by machi1990 — CHANGES_REQUESTED** [review]
Really good change and goes in line with a similar improvement (cache model computation during plugin execution) I'd had in my mind. Thanks for adding this. I've added some minor comments.

File: packages/graphback-core/src/plugin/GraphbackCoreMetadata.ts
Comment (craicoverflow) [reply packages/graphback-core/src/plugin/GraphbackCoreMetadata.ts]: I was not sure what to call it, because it is still technically setting `this.models`. I've renamed it to `buildAndSetModelDefinitions`. Is that suitable?

File: packages/graphback-runtime-knex/tests/data/KnexDbDataProviderTest.ts
Comment (craicoverflow) [reply packages/graphback-runtime-knex/tests/data/KnexDbDataProviderTest.ts:8]: I tried importing the actual package, but got this error, which is isolated to the test files, not sure why it is happening

```shell
Type 'SchemaCRUDPlugin' is not assignable to type 'GraphbackPlugin'.
  Types of property 'transformSchema' are incompatible.
    Type '(metadata: import("/home/ephelan/code/aerogear/graphback/packages/graphback-runtime-knex/node_modules/@graphback/codegen-schema/node_modules/@graphback/core/types/plugin/GraphbackCoreMetadata").GraphbackCoreMetadata) => import("/home/ephelan/code/aerogear/graphback/node_modules/graphql/type/schema").GraphQLSchema' is not assignable to type '(metadata: import("/home/ephelan/code/aerogear/graphback/packages/graphback-core/types/plugin/GraphbackCoreMetadata").GraphbackCoreMetadata) => import("/home/ephelan/code/aerogear/graphback/node_modules/graphql/type/schema").GraphQLSchema'.
      Types of parameters 'metadata' and 'metadata' are incompatible.
        Type 'import("/home/ephelan/code/aerogear/graphback/packages/graphback-core/types/plugin/GraphbackCoreMetadata").GraphbackCoreMetadata' is not assignable to type 'import("/home/ephelan/code/aerogear/graphback/packages/graphback-runtime-knex/node_modules/@graphback/codegen-schema/node_modules/@graphback/core/types/plugin/GraphbackCoreMetadata").GraphbackCoreMetadata'.
          Types have separate declarations of a private property 'supportedCrudMethods'.
```

Comment (machi1990) [reply packages/graphback-runtime-knex/tests/data/KnexDbDataProviderTest.ts:8]: Ah I would not expect that error. Anyway, we can ignore this comments to not derail us from getting this in. 

Maybe we can open a separate issue for this?

Comment (craicoverflow) [reply packages/graphback-runtime-knex/tests/data/KnexDbDataProviderTest.ts]: Yeah, and when I added it as a devDependency I got a circular dependency warning, so that is possibly why:

```
lerna WARN ECYCLE @graphback/codegen-schema -> @graphback/core -> @graphback/codegen-schema
```

Comment (machi1990) [reply packages/graphback-runtime-knex/tests/data/KnexDbDataProviderTest.ts]: Okay, that explains it.

File: packages/graphback-core/src/plugin/GraphbackPluginEngine.ts
Comment (craicoverflow) [reply packages/graphback-core/src/plugin/GraphbackPluginEngine.ts]: I will need to revert to how I had it first, I was seeing integration tests failing in the DataSync plugin (https://github.com/aerogear/graphback/pull/2057/checks?check_run_id=1127868628)

The reason for this is because the DataSync plugin programatically applies extra metadata to the schema, so the model definitions need to be built again _after_ the `transformSchema` hook.

**Comment by wtrocki** [comment]
Can we approve it but wait for the merge. If we release 1.0 in the coming days (early next week) it might mess things up. Just curious if that is something that really needs to be done now considering 1.0 release etc.

**Comment by craicoverflow** [comment]
That is probably the wise choice @wtrocki - this looked to be a simple change but it is having some side effects. I will get it into a "ready" state but close it so we can restore it when ready.

**Comment by craicoverflow** [comment]
RelationshipBuilder blocking this from working fully:

```
packages/graphback-core/tests/MetadataTest.ts
  ● Model has default metadata configuration

    expect(received).toHaveLength(expected)

    Expected length: 1
    Received length: 2
    Received array:  [{"kind": "oneToMany", "owner": "Note", "ownerField": {"args": [{"astNode": {"defaultValue": undefined, "description": undefined, "directives": undefined, "kind": "InputValueDefinition", "name": {"kind": "Name", "value": "filter"}, "type": {"kind": "NamedType", "name": {"kind": "Name", "value": "CommentFilter"}}}, "defaultValue": undefined, "description": null, "name": "filter", "type": "CommentFilter"}], "astNode": {"arguments": [{"defaultValue": undefined, "description": undefined, "directives": undefined, "kind": "InputValueDefinition", "name": {"kind": "Name", "value": "filter"}, "type": {"kind": "NamedType", "name": {"kind": "Name", "value": "CommentFilter"}}}], "description": undefined, "directives": undefined, "kind": "FieldDefinition", "name": {"kind": "Name", "value": "comments"}, "type": {"kind": "ListType", "type": {"kind": "NamedType", "name": {"kind": "Name", "value": "Comment"}}}}, "deprecationReason": undefined, "description": "@oneToMany(field: 'note', key: 'noteId')
    @oneToMany(field: 'note')", "extensions": {"directives": []}, "isDeprecated": false, "name": "comments", "resolve": undefined, "subscribe": undefined, "type": "[Comment]"}, "relationFieldName": "note", "relationForeignKey": "noteId", "relationType": "Comment"}, {"kind": "oneToMany", "owner": "Note", "ownerField": {"args": [{"astNode": {"defaultValue": undefined, "description": undefined, "directives": undefined, "kind": "InputValueDefinition", "name": {"kind": "Name", "value": "filter"}, "type": {"kind": "NamedType", "name": {"kind": "Name", "value": "CommentFilter"}}}, "defaultValue": undefined, "description": null, "name": "filter", "type": "CommentFilter"}], "astNode": {"arguments": [{"defaultValue": undefined, "description": undefined, "directives": undefined, "kind": "InputValueDefinition", "name": {"kind": "Name", "value": "filter"}, "type": {"kind": "NamedType", "name": {"kind": "Name", "value": "CommentFilter"}}}], "description": undefined, "directives": undefined, "kind": "FieldDefinition", "name": {"kind": "Name", "value": "comments"}, "type": {"kind": "ListType", "type": {"kind": "NamedType", "name": {"kind": "Name", "value": "Comment"}}}}, "deprecationReason": undefined, "description": "@oneToMany(field: 'note', key: 'noteId')
    @oneToMany(field: 'note')", "extensions": {"directives": []}, "isDeprecated": false, "name": "comments", "resolve": undefined, "subscribe": undefined, "type": "[Comment]"}, "relationFieldName": "note", "relationForeignKey": "noteId", "relationType": "Comment"}]
```

https://github.com/aerogear/graphback/issues/2023 will need to be completed before merging this. Closing for now.

**Pull request closed without merging**
